packages/x-data-grid/src/hooks/features/rowSelection/useGridRowSelection.ts

```diff
@@ -219,24 +230,72 @@ export const useGridRowSelection = (
       if (resetSelection) {
         logger.debug(`Setting selection for row ${id}`);
 
-        apiRef.current.setRowSelectionModel(isSelected ? [id] : []);
+        const newSelection: GridRowId[] = [];
+        if (isSelected) {
+          newSelection.push(id);
+          if (applyAutoSelection) {
+            const rowsToSelect = findRowsToSelect(
+              apiRef,
+              tree,
+              id,
+              props.rowSelectionPropagation?.descendants ?? false,
+              props.rowSelectionPropagation?.parents ?? false,
+            );
+            rowsToSelect.forEach((rowId) => {
+              newSelection.push(rowId);
+            });
```

Comment:
Side note: there might be some space for optimization here.
For example, `findRowsToSelect` can accept a callback that will be called every time an item is added to `rowsToSelect`. It will make the `forEach` + `push` here unnecessary. Also, `Array.from` in `findRowsToSelect` won't be needed anymore.